extmod/machnie_i2c: Add support for the addrsize parameter

deshipu opened the pull request:
The memory read/write i2c functions now take an optional keyword-only
parameter that specifies the number of bits in the memory address.
The actual transmission is rounded to 8 bits.

Due to the integer type used for the address, for values larger than 16
bits, only 16 bits of address will be sent, and the rest will be padded
with 0s. We can use a larger integer, but we have to decide on the upper
bound. Right now no exception is raised when that happens. For values
smaller than 8, no address is sent. Also no exception then.

Tested with a VL6180 sensor, which has 16-bit register addresses.

This patch makes the resulting binary 84 bytes smaller on the ESP8266.

deshipu committed:
extmod/machnie_i2c: Add support for the addrsize parameter

The memory read/write i2c functions now take an optional keyword-only
parameter that specifies the number of bits in the memory address.
The actual transmission is rounded to 8 bits.

Due to the integer type used for the address, for values larger than 32
bits, only 32 bits of address will be sent, and the rest will be padded
with 0s. Right now no exception is raised when that happens. For values
smaller than 8, no address is sent. Also no exception then.

Tested with a VL6180 sensor, which has 16-bit register addresses.

deshipu commented:
I had to add a condition in there, that unfortunately makes the binary 32 bytes larger again :-(

dpgeorge commented on extmod/machine_i2c.c:
```diff
@@ -415,7 +412,7 @@ STATIC mp_obj_t machine_i2c_writeto_mem(size_t n_args, const mp_obj_t *pos_args,
         { MP_QSTR_memaddr, MP_ARG_REQUIRED | MP_ARG_INT, {.u_int = 0} },
         { MP_QSTR_buf,     MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
-        //{ MP_QSTR_addrsize, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 8} }, TODO
+        { MP_QSTR_addrsize, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 8} },
```
You might be able to reuse the allowed_args array with the readfrom_mem_into function.

deshipu replied:
I'm not sure what you mean. Make it a file-scope global?

dpgeorge replied:
yes

dpgeorge commented:
> We can use a larger integer

Changing the memaddr type from uint16_t to uint32_t should have no effect on code size (might even decrease it).  Then at least it supports up to 32 bits address size.  That should be plenty.

deshipu commented:
I changed the memaddr to be 32 bits, and moved the allowed_args to be a global. I don't have a device that would take 32 bit addresses (I guess I could write a slave like that for the arduino), but on the logic analyzer it looks correct.

By the way, the order of bytes in the address is right now fixed. I assume that for devices that use the reverse order, the user can just convert the addresses themselves before passing to this function.

I also updated the docs.

deshipu commented:
Rebased.

dpgeorge commented on extmod/machine_i2c.c:
```diff
@@ -172,33 +154,26 @@ STATIC int mp_hal_i2c_read_byte(machine_i2c_obj_t *self, uint8_t *val, int nack)
-        if (!mp_hal_i2c_read_byte(self, data++, len == 0)) {
-            goto er;
+    for (int16_t i = addrsize; i > 0; i -= 8) {
+        if (!mp_hal_i2c_write_byte(self, (memaddr >> (i - 8)) & 0xff)) {
```
If addrsize < 8 then funny things will happen with the `>>`.  You can probably make this loop simpler by doing something like `for (int i = addrsize >> 3; i-- > 0;) { if (!mp_hal_i2c_write_byte(self, memaddr >> i)) { ...`.  Note that you don't need to mask `&0xff` because write_byte can't send more than 8 bits.

deshipu replied:
But I need to shift the memaddr right by 8 bits on every iteration, what you propose only shifts it by one? I would need (memaddr >> (i << 3))?

deshipu replied:
I fixed the loop to not require the subtraction on every iteration, and to stop as soon as the counter is negative. I also added rounding the `addrsize` to a multiple of 8. This should solve the problems.

dpgeorge commented on extmod/machine_i2c.c:
```diff
@@ -373,7 +368,7 @@ STATIC mp_obj_t machine_i2c_readfrom_mem(size_t n_args, const mp_obj_t *pos_args
         { MP_QSTR_memaddr, MP_ARG_REQUIRED | MP_ARG_INT, {.u_int = 0} },
         { MP_QSTR_n,       MP_ARG_REQUIRED | MP_ARG_INT, {.u_int = 0} },
-        //{ MP_QSTR_addrsize, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 8} }, TODO
+        { MP_QSTR_addrsize, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 8} },
```
You could remove this arg array and use the common machine_i2c_mem_allowed_args array instead.  The only difference would be that the `n` arg is now called `buf`.  That could be renamed to simply `arg` so as not to be confusing.

deshipu replied:
Done.

dpgeorge commented on extmod/machine_i2c.c:
```diff
@@ -384,51 +379,53 @@ STATIC mp_obj_t machine_i2c_readfrom_mem(size_t n_args, const mp_obj_t *pos_args
     vstr_init_len(&vstr, args[ARG_n].u_int);
```
This would need to become `mp_obj_get_int(args[ARG_n].u_obj)` if a common arg array is used.

deshipu replied:
Done.

dpgeorge commented on extmod/machine_i2c.c:
```diff
@@ -218,27 +193,47 @@ STATIC void mp_hal_i2c_write_mem(machine_i2c_obj_t *self, uint8_t addr, uint16_t
+}
+
+STATIC void mp_hal_i2c_read(machine_i2c_obj_t *self, uint8_t addr,
+        uint8_t *dest, size_t len) {
```
This line doesn't need to be broken.  Line limit is roughly 100 chars.

dpgeorge commented:
Merged in 219245e10fa01ab81a7dd10b64fffd07e2ec1a49, thanks.

I changed the order of parameters in some C funcs so that memaddr and addrsize are adjacent.  I also removed the rounding of the addrsize to 8, since that's not necessary to get correct behaviour, and passing an addrsize that's not a multiple of 8 should anyway be considered undefined behaviour.

deshipu commented:
Thank you. Do you think it would make sense for me to also change i2c.scan to use writes?

dpgeorge commented:
> Do you think it would make sense for me to also change i2c.scan to use writes?

If you can test it on a few devices that would be good.  I looked at Arduino examples for scan and they use empty writes.
